{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def base_Capitalizer(seqList):\n",
    "    uppercaseSeqList = []\n",
    "    dnaChecker = True\n",
    "    for nucleotide in seqList:\n",
    "        if nucleotide.casefold() in ('a', 'c', 'g', 't'):\n",
    "            uppercaseSeqList.append(nucleotide.upper())\n",
    "        else:\n",
    "            dnaChecker = False\n",
    "    if dnaChecker == True:\n",
    "        seqList = uppercaseSeqList\n",
    "    else:\n",
    "        return \"Warning: This is not a DNA sequence.\"\n",
    "    return seqList\n",
    "\n",
    "def dna_compSeqMaker(dnaSeqList):\n",
    "    three_to_fivecompSeqList = []\n",
    "    for nucleotide in dnaSeqList:\n",
    "        if nucleotide.casefold() == 't':\n",
    "            three_to_fivecompSeqList.append('A')\n",
    "                \n",
    "        elif nucleotide.casefold() == 'a':\n",
    "            three_to_fivecompSeqList.append('T')\n",
    "                \n",
    "        elif nucleotide.casefold() == 'c':\n",
    "            three_to_fivecompSeqList.append('G')\n",
    "           \n",
    "        elif nucleotide.casefold() == 'g':\n",
    "            three_to_fivecompSeqList.append('C')\n",
    "        else:\n",
    "            dnaChecker = False\n",
    "            break           \n",
    "    compSeqList = three_to_fivecompSeqList[::-1]\n",
    "    return compSeqList\n",
    "        \n",
    "def dna_startCodonFinder(seqList, positionToStart):\n",
    "    counter = positionToStart\n",
    "    found = False\n",
    "    while counter <= len(seqList)-1 and found is False:\n",
    "        if counter + 2 <= len(seqList)-1 and seqList[counter : counter + 3] == ['A', 'T', 'G']:\n",
    "            found = True\n",
    "        else: \n",
    "            counter += 1      \n",
    "    if found == True:\n",
    "        return counter + 2\n",
    "    else:\n",
    "        return \"No start codon was found\"\n",
    "\n",
    "def dna_orfFinder(dnaSeqList, positionToStart):\n",
    "    thirdStartBase_position = dna_startCodonFinder(dnaSeqList, positionToStart)\n",
    "    if type(thirdStartBase_position) == str:\n",
    "        return thirdStartBase_position\n",
    "    else: \n",
    "        counter = thirdStartBase_position + 1 \n",
    "        found = False\n",
    "        while counter <= len(dnaSeqList)-1 and found is False:\n",
    "            if counter + 2 <= len(dnaSeqList)-1 and dnaSeqList[counter : counter + 3] in (['T', 'A', 'A'], ['T', 'A', 'G'], ['T', 'G', 'A']):\n",
    "                found = True\n",
    "            else:\n",
    "                counter += 3\n",
    "        if found == True:\n",
    "            return [thirdStartBase_position - 2, counter + 2]\n",
    "        else:\n",
    "            return \"No stop codon was found\"\n",
    "\n",
    "def dna_wholeSeqOrfFinder(dnaSeqList):\n",
    "    positionToStart = 0\n",
    "    orfsInSeq = {}\n",
    "    orfLabel = 'ORF '\n",
    "    orfCounter = 0\n",
    "    while positionToStart <= len(dnaSeqList)-1:\n",
    "        orf = dna_orfFinder(dnaSeqList, positionToStart)\n",
    "        if type(orf) == list:\n",
    "            orfCounter += 1\n",
    "            orfsInSeq[orfLabel + str(orfCounter)] = orf\n",
    "            positionToStart = orf[1] + 1\n",
    "        else:\n",
    "            if len(orfsInSeq) == 0:\n",
    "                return orf\n",
    "            else:\n",
    "                break\n",
    "    if len(orfsInSeq) > 1:\n",
    "        orderedOrfs = {j: i for j, i in sorted(orfsInSeq.items(), key=lambda item: abs(item[1][0]-item[1][1]), reverse=True)}\n",
    "        return orderedOrfs\n",
    "    else:\n",
    "        return orfsInSeq\n",
    "    \n",
    "    \n",
    "def dna_orfPrinter(orfOutputDict):\n",
    "    if type(orfOutputDict) != dict:\n",
    "        return orfOutputDict \n",
    "    else:\n",
    "        if len(orfOutputDict) == 1:\n",
    "            printStatement = 'There is only one open reading frame in this sequence which spans from base ' + str(orfOutputDict['ORF 1'][0]) + ' to ' + str(orfOutputDict['ORF 1'][1]) + ' making it ' + str(orfOutputDict['ORF 1'][1] - orfOutputDict['ORF 1'][0]) + ' bases long'\n",
    "            return printStatement\n",
    "        else:\n",
    "            printStatement = 'The following ' + str(len(orfOutputDict)) + ' open reading frames in the sequence are ordered from greatest to least in size: ' '\\n'\n",
    "            for key, (start, stop) in orfOutputDict.items():\n",
    "                printStatement += '{}: {} bases long, spanning from bases {} to {}\\n'.format(key, stop-start + 1, start, stop)\n",
    "            return printStatement\n",
    "    \n",
    "def rna_codingToMRNA(dnaSeqList):\n",
    "    mrnaList = []\n",
    "    rnaChecker = True\n",
    "    for nucleotide in dnaSeqList:\n",
    "        if nucleotide.casefold() in (\"a\", \"c\", \"g\"):\n",
    "            mrnaList.append(nucleotide.upper()) \n",
    "        elif nucleotide.casefold() == 't':\n",
    "            mrnaList.append('U') \n",
    "        else:\n",
    "            if nucleotide.casefold() not in  (\"a\", \"t\", \"c\", \"g\"):\n",
    "                rnaChecker = False\n",
    "                break              \n",
    "    if rnaChecker == True: \n",
    "        return mrnaList\n",
    "    else: \n",
    "        return \"Warning: This is not a DNA sequence.\" \n",
    "\n",
    "def mRNA_rnaToAminoAcidSeq(list):\n",
    "    counter = 0\n",
    "    aminoAcidList = []\n",
    "    while counter <= len(list) - 1:\n",
    "        if counter == 0:\n",
    "            aminoAcidList.extend(['Nter—', 'M'])\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['U', 'U', 'U'], ['U', 'U', 'C']):\n",
    "            aminoAcidList.append('F')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['U', 'U', 'A'], ['U', 'U', 'G'], ['C', 'U', 'U'], ['C', 'U', 'C'], ['C', 'U', 'A'], ['C', 'U', 'G']):\n",
    "            aminoAcidList.append('L')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['A', 'U', 'U'], ['A', 'U', 'C'], ['A', 'U', 'A']):\n",
    "            aminoAcidList.append('I')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] == ['A', 'U', 'G']:\n",
    "            aminoAcidList.append('M')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['G', 'U', 'U'], ['G', 'U', 'C'], ['G', 'U', 'A'], ['G', 'U', 'G']):\n",
    "            aminoAcidList.append('V')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['U', 'C', 'U'], ['U', 'C', 'C'], ['U', 'C', 'A'], ['U', 'C', 'G'], ['A', 'G', 'U'], ['A', 'G', 'C']):\n",
    "            aminoAcidList.append('S')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['C', 'C', 'U'], ['C', 'C', 'C'], ['C', 'C', 'A'], ['C', 'C', 'G']):\n",
    "            aminoAcidList.append('P')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['A', 'C', 'U'], ['A', 'C', 'C'], ['A', 'C', 'A'], ['A', 'C', 'G']):\n",
    "            aminoAcidList.append('T')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['G', 'C', 'U'], ['G', 'C', 'C'], ['G', 'C', 'A'], ['G', 'C', 'G']):\n",
    "            aminoAcidList.append('A')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['U', 'A', 'U'], ['U', 'A', 'C']):\n",
    "            aminoAcidList.append('Y')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['C', 'A', 'U'], ['C', 'A', 'C']):\n",
    "            aminoAcidList.append('H')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['C', 'A', 'A'], ['C', 'A', 'G']):\n",
    "            aminoAcidList.append('Q')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['A', 'A', 'U'], ['A', 'A', 'C']):\n",
    "            aminoAcidList.append('N')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['A', 'A', 'A'], ['A', 'A', 'G']):\n",
    "            aminoAcidList.append('K')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['G', 'A', 'U'], ['G', 'A', 'C']):\n",
    "            aminoAcidList.append('D')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['G', 'A', 'A'], ['G', 'A', 'G']):\n",
    "            aminoAcidList.append('E')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['U', 'G', 'U'], ['U', 'G', 'C']):\n",
    "            aminoAcidList.append('C')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] == ['U', 'G', 'G']:\n",
    "            aminoAcidList.append('W')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['C', 'G', 'U'], ['C', 'G', 'C'], ['C', 'G', 'A'], ['C', 'G', 'G'], ['A', 'G', 'A'], ['A', 'G', 'G']):\n",
    "            aminoAcidList.append('R')\n",
    "            counter += 3\n",
    "        elif list[counter : counter + 3] in (['G', 'G', 'U'], ['G', 'G', 'C'], ['G', 'G', 'A'], ['G', 'G', 'G']):\n",
    "            aminoAcidList.append('G')\n",
    "            counter += 3            \n",
    "        elif list[counter : counter + 3] in (['U', 'A', 'A'], ['U', 'A', 'G'], ['U', 'G', 'A']):\n",
    "                aminoAcidList.append('—Cter')\n",
    "                counter += 3\n",
    "        else:\n",
    "            aminoAcidList.append('ERROR')\n",
    "            counter += 3\n",
    "    return aminoAcidList     \n",
    "\n",
    "def protein_printer(proteinOutputDict):\n",
    "    if type(proteinOutputDict) != dict:\n",
    "        return 'Warning: This is not the proper input.'\n",
    "    else:\n",
    "        if len(proteinOutputDict) == 1:\n",
    "            printStatement = 'There is only one protein in the used sequence which is ' + str(len(proteinOutputDict[key]) - 2) + ' bases long'\n",
    "            return printStatement\n",
    "        else:\n",
    "            printStatement = 'The following ' + str(len(proteinOutputDict)) + ' proteins in the sequence are ordered from greatest to least in size: ' '\\n'\n",
    "            for key, value in proteinOutputDict.items():\n",
    "                printStatement += '{}: {} amino acids long\\n'.format(key, len(value) - 2)\n",
    "            return printStatement\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
